shell/platform/android/io/flutter/embedding/android/FlutterFragmentActivity.java

```diff
@@ -0,0 +1,671 @@
+      return desiredDartEntrypoint != null ? desiredDartEntrypoint : DEFAULT_DART_ENTRYPOINT;
+    } catch (PackageManager.NameNotFoundException e) {
+      return DEFAULT_DART_ENTRYPOINT;
+    }
+  }
+
+  /**
+   * The initial route that a Flutter app will render upon loading and executing its Dart code.
+   * <p>
+   * This preference can be controlled with 2 methods:
+   * <ol>
+   *   <li>Pass a boolean as {@link FlutterActivityLaunchConfigs#EXTRA_INITIAL_ROUTE} with the
+   *     launching {@code Intent}, or</li>
+   *   <li>Set a {@code <meta-data>} called
+   *     {@link FlutterActivityLaunchConfigs#INITIAL_ROUTE_META_DATA_KEY} for this {@code Activity}
+   *     in the Android manifest.</li>
+   * </ol>
+   * If both preferences are set, the {@code Intent} preference takes priority.
+   * <p>
+   * The reason that a {@code <meta-data>} preference is supported is because this {@code Activity}
+   * might be the very first {@code Activity} launched, which means the developer won't have
+   * control over the incoming {@code Intent}.
+   * <p>
+   * Subclasses may override this method to directly control the initial route.
+   */
+  @NonNull
+  protected String getInitialRoute() {
+    if (getIntent().hasExtra(EXTRA_INITIAL_ROUTE)) {
+      return getIntent().getStringExtra(EXTRA_INITIAL_ROUTE);
+    }
+
+    try {
+      ActivityInfo activityInfo = getPackageManager().getActivityInfo(
+          getComponentName(),
+          PackageManager.GET_META_DATA|PackageManager.GET_ACTIVITIES
+      );
+      Bundle metadata = activityInfo.metaData;
+      String desiredInitialRoute = metadata != null ? metadata.getString(INITIAL_ROUTE_META_DATA_KEY) : null;
+      return desiredInitialRoute != null ? desiredInitialRoute : DEFAULT_INITIAL_ROUTE;
+    } catch (PackageManager.NameNotFoundException e) {
+      return DEFAULT_INITIAL_ROUTE;
+    }
+  }
+
+  /**
+   * Returns the ID of a statically cached {@link FlutterEngine} to use within this
+   * {@code FlutterFragmentActivity}, or {@code null} if this {@code FlutterFragmentActivity} does not want
+   * to use a cached {@link FlutterEngine}.
+   */
+  @Nullable
+  protected String getCachedEngineId() {
+    return getIntent().getStringExtra(EXTRA_CACHED_ENGINE_ID);
+  }
+
+  /**
+   * The desired window background mode of this {@code Activity}, which defaults to
+   * {@link BackgroundMode#opaque}.
+   */
+  @NonNull
+  protected BackgroundMode getBackgroundMode() {
+    if (getIntent().hasExtra(EXTRA_BACKGROUND_MODE)) {
+      return BackgroundMode.valueOf(getIntent().getStringExtra(EXTRA_BACKGROUND_MODE));
+    } else {
+      return BackgroundMode.opaque;
+    }
+  }
+
+  /**
+   * Returns true if Flutter is running in "debug mode", and false otherwise.
+   * <p>
+   * Debug mode allows Flutter to operate with hot reload and hot restart. Release mode does not.
+   */
+  private boolean isDebuggable() {
+    return (getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;
+  }
```

Comment:
It seems like most of the code is duplicated. Could we extract somehow?

Reply:
It is possible to extract, but I'm not sure we should. As with the delegate change that we made, I'm hesitant to tear apart Activity implementations into other objects that do all the work on behalf of the Activity. When that happens, reading the Activity becomes an exercise in constantly jumping to other files and back. Do you think it's a meaningful duplication risk to leave this as-is?

Reply:
Discussed offline. You decide. I don't feel strongly, as long as we leave hints for other maintainers.

Reply:
We spoke offline. I'm suggesting that we leave duplicated for now for readability. If we find that this creates maintenance problems in the future then we can re-evaluate. I'll post a PR immediately after this one that adds javadoc comments mentioning that these files need to change together.